tools/storage.py

```diff
@@ -0,0 +1,173 @@
+        return self.header.base_offset + self.header.record_count - 1
+
+    def _crc_header_be_bytes(self):
+        # encode header back to big-endian for crc calculation
+        return struct.pack(">" + HDR_FMT_CRC, *self.header[5:])
+
+    @staticmethod
+    def from_file(f, index):
+        data = f.read(HEADER_SIZE)
+        if len(data) == HEADER_SIZE:
+            header = Header(*struct.unpack(HDR_FMT_RP, data))
+            # it appears that we may have hit a truncation point if all of the
+            # fields in the header are zeros
+            if all(map(lambda v: v == 0, header)):
+                return
+            records_size = header.batch_size - HEADER_SIZE
+            data = f.read(records_size)
+            assert len(data) == records_size
+            return Batch(index, header, data)
+        assert len(data) == 0
+
+
+class Segment:
+    def __init__(self, path):
+        self.path = path
+        self.__read_batches()
+
+    def __read_batches(self):
+        index = 1
+        with open(self.path, "rb") as f:
+            while True:
+                batch = Batch.from_file(f, index)
+                if not batch:
+                    break
+                index += 1
+
+    def dump(self):
+        if self.batches:
+            next_offset = self.batches[0].header.base_offset
+        index = 0
+        for batch in self.batches:
+            if index < 3 or index == (len(self.batches) - 1):
+                logger.info(batch.header.base_offset)
+            if index == 3 and len(self.batches) > 4:
+                logger.info("...")
+            if batch.header.base_offset != next_offset:
+                logger.info("hole discovered at offset {} expected {}".format(
+                    batch.header.base_offset, next_offset))
+                break
+            next_offset = batch.last_offset() + 1
+            index += 1
+
+
+class Ntp:
+    def __init__(self, base_dir, namespace, topic, partition):
+        self.base_dir = base_dir
+        self.nspace = namespace
+        self.topic = topic
+        self.partition = partition
+        self.path = os.path.join(self.base_dir, self.nspace, self.topic,
+                                 str(self.partition))
+        pattern = os.path.join(self.path, "*.log")
+        self.segments = glob.iglob(pattern)
+
+    def __str__(self):
+        return "{0.nspace}/{0.topic}/{0.partition}".format(self)
+
+
+class Store:
+    def __init__(self, base_dir):
+        self.base_dir = os.path.abspath(base_dir)
+        self.ntps = []
+        self.__search()
+
+    def __search(self):
+        dirs = os.walk(self.base_dir)
+        for ntpd in (p[0] for p in dirs if not p[1]):
+            head, part = os.path.split(ntpd)
+            head, topic = os.path.split(head)
+            head, nspace = os.path.split(head)
+            assert head == self.base_dir
+            ntp = Ntp(self.base_dir, nspace, topic, int(part))
+            self.ntps.append(ntp)
+
+def main():
```

Comment:
Since this is a move from another repo theres no diff to observe. But I can attest that the only major change here was the lines below this comment. Above this comment I added, import os, sys, & logger, and modified all calls to print() to use logger.info/error

Reply:
can't you import the history of this file into the tree?

Reply:
This was a great idea as it revealed a bug , i did not import the most recent change of that file. Change has been split into two, 1. the mv across repos and 2. the edits i've made.